tensorflow/contrib/layers/python/layers/layers.py

```diff
@@ -2129,6 +2131,97 @@ def legacy_fully_connected(x,
 
     return _apply_activation(y, activation_fn, output_collections)
 
+def adaptive_softmax_loss(inputs,
+                          labels,
+                          cutoff,
+                          project_factor=4,
+                          initializer=None,
+                          name=None):
+  """Computes and returns the adaptive softmax loss (a improvement of 
+  hierarchical softmax).
+    
+  See [Efficient softmax approximation for GPUs](https://arxiv.org/pdf/1609.04309v2.pdf).
+        
+  This is a faster way to train a softmax classifier over a huge number of 
+  classes, and can be used for **both training and prediction**. For example, it 
+  can be used for training a Language Model with a very huge vocabulary, and 
+  the trained languaed model can be used in speech recognition, text generation, 
+  and machine translation very efficiently.
+  
+  Args:
+    inputs: A `Tensor` of shape `[batch_size, dim]`.  The forward
+      activations of the input network.
+    labels: `Tensor` of shape `[d_0, d_1, ..., d_{r-2}]` and dtype `int32` or
+      `int64`. Each entry in `labels` must be an index in `[0, num_classes)`.
+    cutoff: A list indicating the limits of the different clusters.
+    project_factor: A floating point value greater or equal to 1.0. The projection 
+      factor between two neighboring clusters.
+    initializer: Initializer for adaptive softmax variables (optional).
+    name: A name for the operation (optional).
+
+  Returns:
+    loss: A `batch_size` 1-D tensor of the adaptive softmax cross entropy loss.
+    training_losses: A list of 1-D tensors of adaptive softmax loss for each 
+      cluster, which can be used for calculating the gradients and back 
+      propagation when training.
+  """
+  input_dim = int(inputs.get_shape()[1])
```

Comment:
this will fail if inputs.get_shape().ndims is None or if get_shape()[1] is None.  same for the line below.

is it critical that these values are known at graph build time, or can you use TF ops here?